StreamChatClient_v3/ChatClient.swift

```diff
@@ -73,64 +67,61 @@ public class Client<ExtraData: ExtraDataTypes> {
     private(set) var backgroundWorkers: [Worker]!
     
     /// The `APIClient` instance `Client` uses to communicate with Stream REST API.
-    private(set) lazy var apiClient: APIClient = {
-        let apiClient = self.environment
-            .apiClientBuilder(self.config.apiKey, config.baseURL.restAPIBaseURL, self.urlSessionConfiguration)
-        self.webSocketClient.connectionStateDelegate = apiClient
+    lazy var apiClient: APIClient = {
+        var encoder = environment.requestEncoderBuilder(config.baseURL.restAPIBaseURL, config.apiKey)
+        encoder.connectionDetailsProviderDelegate = self
+        
+        let decoder = environment.requestDecoderBuilder()
+        
+        let apiClient = environment.apiClientBuilder(urlSessionConfiguration, encoder, decoder)
         return apiClient
     }()
     
     /// The `WebSocketClient` instance `Client` uses to communicate with Stream WS servers.
-    private(set) lazy var webSocketClient: WebSocketClient = {
-        // Set up event middlewares
-        let middlewares: [EventMiddleware] = [
-            // TODO: Add more middlewares
-            EventDataProcessorMiddleware<ExtraData>(database: self.persistentContainer),
-            HealthCheckFilter()
+    lazy var webSocketClient: WebSocketClient = {
+        // Create a connection request
+        let webSocketEndpoint = webSocketConnectEndpoint(userId: self.currentUserId)
+        
+        var encoder = environment.requestEncoderBuilder(config.baseURL.webSocketBaseURL, config.apiKey)
+        encoder.connectionDetailsProviderDelegate = self
+        
+        let connectEndpoint = webSocketConnectEndpoint(userId: currentUserId)
+        
+        // TODO: Add more middlewares
+        let eventMiddlewares = [
```

Comment:
If the `webSocketClient` is lazy var we can move this outside for easier management

Reply:
Hmm, OK, why not 🤷‍♂️